### PR TITLE
Logan/convnext testing

### PR DESCRIPTION
**Pull request by Llechuga03**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

**WARNING ⚠️** this PR is very large, summary may not cover all changes.

### 🌟 Summary
Experimental integration of a ConvNeXt-Tiny backbone into YOLOv5 with HPC-specific training artifacts and notebook output updates, but with several environment-specific and debugging changes that are unlikely to be merge-ready.

### 📊 Key Changes
- 🧠 **New ConvNeXt backbone modules**
  - Adds `ConvNextTiny`, `ConvNextP3`, and `ConvNextP4` in `models/common.py`, using `timm.create_model('convnext_tiny', features_only=True, out_indices=[1, 2, 3])`.
  - Implements a global `_CONVNEXT_FEATURES` registry to pass multi-scale P3/P4/P5 features between modules.
  - Adapters map ConvNeXt feature channels to YOLOv5-style heads (P3→128, P4→256, P5→`c2`).
- 🧩 **Model parser integration**
  - Registers `ConvNextTiny`, `ConvNextP3`, and `ConvNextP4` in `models/yolo.py::parse_model` with fixed output channel assumptions.
  - Adds verbose `print(f"Module: {m.__name__}, args: {args}")` for every parsed layer (debug output).
- 🧷 **Dataset & environment coupling**
  - Replaces the generic `data/coco.yaml` path with a hard-coded absolute HPC path: `/projectnb/ec523bn/projects/ConvNeXt-YOLO/datasets/coco`.
  - Introduces `train.qsub` for a specific SGE/HPC setup and adds `logs/yolo.qlog` as a tracked log file.
- 📓 **Notebook output updates**
  - Modifies `tutorial.ipynb` to include executed cell outputs, environment-specific paths, warnings, and minor code/comment tweaks (e.g., commented-out clone, added `print(sys.executable)`).
- 📝 **Minor metadata/comment tweaks**
  - Adjusts a comment in `models/yolov5s.yaml` to annotate backbone layer count.

### 🎯 Purpose & Impact
- 🎯 **Intended purpose**
  - Enable experimentation with a ConvNeXt-Tiny backbone inside YOLOv5 for potential accuracy/feature improvements on COCO.
  - Capture and reuse multi-scale ConvNeXt features without modifying the standard detection head logic.
- ⚠️ **Potential issues / merge blockers**
  - Hard-coded absolute dataset path in `coco.yaml` breaks portability and standard usage patterns.
  - New runtime dependency on `timm` is added without updates to `requirements.txt` or documentation.
  - Global `_CONVNEXT_FEATURES` state and input `data_ptr()` keys are fragile for multi-GPU, dataloader workers, or concurrent inference.
  - Extensive `print` statements (initialization, stage channels, `parse_model` logs, and warning fallbacks) will spam stdout and slow training/CI.
  - `logs/yolo.qlog`, `train.qsub`, and executed `tutorial.ipynb` outputs are project- and environment-specific artifacts that generally should not be committed.
- 📈 **User impact (if refined)**
  - A cleaned-up version could offer users an additional ConvNeXt-based backbone option for detection experiments.
  - Proper integration would require configuration examples, docs, CI coverage, and robust handling of features without global mutable state.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Llechuga03, thank you for submitting a `ultralytics/yolov5` 🚀 PR! This is an automated review assistant; an Ultralytics engineer will also review your changes.

To help ensure a smooth integration of your work, please review the checklist below:

- ✅ **Define a Purpose**: Clearly explain the purpose of your ConvNeXt-related changes and training artifacts in the PR description, and link to any [relevant issues](https://github.com/ultralytics/yolov5/issues). Ensure your commit messages are clear, concise, and follow the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/yolov5` `main` branch. If it is behind, update it by clicking the **Update branch** button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues (for example, by removing environment-specific paths/logs or debug prints if they cause problems).
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features, such as adding a ConvNeXt backbone or new dependencies like `timm`.
- ✅ **Add Tests**: If applicable, include or update tests to cover ConvNeXt integration and ensure that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing `I have read the CLA Document and I sign the CLA` in a new comment on this PR.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your feature or experiment. Consider removing environment-specific artifacts (e.g., cluster job scripts, absolute dataset paths, local logs, and executed notebook outputs) if they are not required for the core ConvNeXt integration.  _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). If you have any questions about how to best integrate ConvNeXt or clean up the PR for merge, feel free to ask in a comment. Thank you for contributing to Ultralytics! 🚀🤝

File: models/common.py [hunk 1]
```diff
@@ -17,6 +17,7 @@
 import numpy as np
 import pandas as pd
 import requests
+import timm
```
Comment (UltralyticsAssistant) [inline models/common.py:20]: ⚠️ **HIGH**: `timm` is now an unconditional import, but it is not declared in `requirements.txt` and is optional for most users. This will raise `ModuleNotFoundError` whenever someone trains/evaluates without having `timm` installed. Please gate the import (e.g., lazy load with a clear error) or add the dependency to the runtime requirements to avoid breaking default installs.

File: models/common.py [hunk 2]
```diff
@@ -1121,3 +1122,126 @@ def forward(self, x):
...
+        
+        self.p5_adapter = nn.Sequential(
+            nn.Conv2d(stage_channels[2], c2, 1, bias=False),
+            nn.BatchNorm2d(c2),
+            nn.SiLU(inplace=True)
+        )
+        
+    def forward(self, x):
+        """Extract features and store in global registry keyed by batch"""
+        global _CONVNEXT_FEATURES
```
Comment (UltralyticsAssistant) [inline models/common.py:1184]: ⚠️ **HIGH**: ConvNext integration relies on a global `_CONVNEXT_FEATURES` cache keyed only by `x.data_ptr()`, and it is never cleared. In DDP/amp or even sequential dataloader workers, multiple inputs can share the same `data_ptr`, leading to cross-sample feature leakage. The dictionary will also grow without bound, leaking GPU tensors. Please pass multi-scale features through the graph (e.g., return a tuple) or at least scope the cache per forward and clear it after use.

File: data/coco.yaml [hunk 3]
```diff
@@ -8,7 +8,7 @@
 #     └── coco  ← downloads here (20.1 GB)
 
 # Train/val/test sets as 1) dir: path/to/imgs, 2) file: path/to/imgs.txt, or 3) list: [path/to/imgs1, path/to/imgs2, ..]
-path: ../datasets/coco # dataset root dir
+path: /projectnb/ec523bn/projects/ConvNeXt-YOLO/datasets/coco # dataset root dir
```
Comment (UltralyticsAssistant) [inline data/coco.yaml:11]: 💡 **MEDIUM**: Hard-coding the dataset `path` to an HPC-specific absolute directory makes the default COCO config unusable for most users (local training, CI, Ultralytics HUB, etc.). Please keep paths relative or configurable so `train.py --data coco.yaml` continues to work out-of-the-box.

**Suggested change:**
```suggestion
path: ../datasets/coco # dataset root dir
```

File: models/common.py [hunk 4]
```diff
@@ -1121,3 +1122,126 @@ def forward(self, x):
...
+
+class ConvNextP4(nn.Module):
+    """
+    Helper module to extract P4 features from ConvNext.
+    """
+    
+    def __init__(self, c1=None, c2=None):
+        super().__init__()
+        
+    def forward(self, x):
```
Comment (UltralyticsAssistant) [inline models/common.py:1214]: 💡 **MEDIUM**: `ConvNextP4`/`ConvNextP3` ignore their input and silently return whatever is cached under `_CONVNEXT_FEATURES['latest']`. When the cache is missing (different batch order, TorchScript, export), the modules just return the incoming tensor even though its shape is wrong for the neck. That hides bugs and yields hard-to-debug size errors downstream; please raise a descriptive exception instead of passing through `x`.

File: models/yolo.py [hunk 5]
```diff
@@ -444,9 +447,17 @@ def parse_model(d, ch):
...
+            # ConvNextTiny: first arg is output channels (c2)
+            c2 = args[0] if args else 512
+        elif m in {ConvNextP3, ConvNextP4}:
+            # These helper modules retrieve stored features
+            # ConvNextP3 outputs 128 channels, ConvNextP4 outputs 256 channels
+            c2 = 128 if m is ConvNextP3 else 256
         else:
             c2 = ch[f]
 
+        print(f"Module: {m.__name__}, args: {args}")
```
Comment (UltralyticsAssistant) [inline models/yolo.py:460]: 📝 **LOW**: `parse_model` now prints every layer (`print(f"Module: {m.__name__}, ... )`). With large configs this produces thousands of log lines and slows script startup. Could you guard this behind a debug flag or remove it before merge to keep default runs quiet and deterministic?

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Main blockers: the new ConvNeXt backbone brings in an unconditional `timm` dependency that will break default installs, and the global `_CONVNEXT_FEATURES` cache is unsafe (feature leakage across batches/workers and unbounded memory growth). Additionally, the new dataset path change hardcodes an HPC-specific directory, the helper heads silently mask cache misses, and `parse_model` now spams stdout. Addressing these issues will make the experiments safer to merge.

💬 Posted 5 inline comments

⚠️ **Large PR**: Review focused on critical issues. Some details may not be covered.
